### PR TITLE
Alternatives: adopt new NN handling alternatives route flow

### PR DESCRIPTION
**Pull request by RingerJK**
### Description
Alternative routes. Adopt new NN handling flow:
- Alternatives observer does not spread routes that has been set via `MapboxNavigation#setRoutes`;
- Alternatives observer does not spread routes if set to `MapboxNavigation#setRoutes`alternatives routes + one or few additional routes.

### CODE REVIEW

File: gradle/dependencies.gradle [hunk 1]
```diff
@@ -13,7 +13,7 @@ ext {
   // version which we should use in this build
   def mapboxNavigatorVersion = System.getenv("FORCE_MAPBOX_NAVIGATION_NATIVE_VERSION")
   if (mapboxNavigatorVersion == null || mapboxNavigatorVersion == '') {
-      mapboxNavigatorVersion = '112.0.0'
+      mapboxNavigatorVersion = 'master-SNAPSHOT'
```
Comment (RingerJK) [inline gradle/dependencies.gradle:16]: will be replaced with approprate version of NN

**Comment by dzinad** [comment]
> Alternatives observer does not spread routes that has been set via MapboxNavigation#setRoutes;
Alternatives observer does not spread routes if set to MapboxNavigation#setRoutesalternatives routes + one or few additional routes.

Do you mean that it was implemented by NN and you just removed the code that we don't need any longer?

File: CHANGELOG.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ Mapbox welcomes participation and contributions from everyone.
 - Added Android 13 support. [#6196](https://github.com/mapbox/mapbox-navigation-android/pull/6196)
 - Declared [POST_NOTIFICATIONS](https://developer.android.com/reference/android/Manifest.permission#POST_NOTIFICATIONS) permission in SDK's AndroidManifest.xml. It is highly recommended for apps to request the permission in runtime. Without it, the SDK will not be able to show the notification with trip progress in the notification drawer for apps that target Android 13 or higher. [#6196](https://github.com/mapbox/mapbox-navigation-android/pull/6196)
 - Marked `MapboxNavigationApp` and `MapboxMavigationObserver` methods with `@UiThread` annotations. [#6254](https://github.com/mapbox/mapbox-navigation-android/pull/6254)
+- Fixed `RouteAlternativesObserver`: does not spread alternative if set to `MapboxNavigation#setRoutes` has been received alternatives plus one or a few additional routes. [#6214](https://github.com/mapbox/mapbox-navigation-android/pull/6214)
```
Comment (LukasPaczos) [inline CHANGELOG.md:20]: Does anything actually change for developers? Weren't our flag (removed in this PR) prevent this from happening, or where there edge cases that we did not consider that are fixed now?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 3]
```diff
@@ -113,11 +116,149 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        countDownLatch.await()
+
         // Verify alternative routes events were triggered.
         assertEquals(2, routes.size)
         assertTrue(mapboxNavigation.getRoutes().isNotEmpty())
         assertNotNull(firstAlternative.routeProgress)
         assertNotNull(firstAlternative.alternatives)
+
+        firstAlternative.alternatives!!.forEach { alternativeRoute ->
+            routes.forEach { originalRoutes ->
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:174]: Maybe `originalRoute`?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 4]
```diff
@@ -113,11 +116,149 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        }
+
+        // The alternative route is missed, so we expect the route to be removed.
+        val firstAlternative = RouteAlternativesIdlingResource(mapboxNavigation)
+        firstAlternative.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("alternatives failed") {
+            Espresso.onIdle()
+        }
+        firstAlternative.unregister()
+        assertTrue(firstAlternative.calledOnMainThread)
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:156]: The test says you want to check that the observer is not triggered. But here you check that the flag is true and it is set to true only when it's triggered. I'm confused.
Why do you need this check at all?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 5]
```diff
@@ -113,11 +116,149 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        countDownLatch.await()
+
+        // Verify alternative routes events were triggered.
+        assertNotNull(firstAlternative.routeProgress)
+        assertNotNull(firstAlternative.alternatives)
+
+        // Verify next alternative routes events were triggered without set routes
+        assertNotNull(nextAlternativeObserver.alternatives)
+        assertNotNull(nextAlternativeObserver.routeProgress)
+        assertTrue(setRoutes.size >= 2)
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:254]: You form that list in the test, is this really what you want to check?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 6]
```diff
@@ -113,11 +116,149 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+            assertEquals(1, it.size)
+        }
+
+        val nextAlternativeObserver = RouteAlternativesIdlingResource(mapboxNavigation)
+
+        var setRoutes: List<DirectionsRoute> = emptyList()
+        runOnMainSync {
+            setRoutes = mutableListOf(
+                mapboxNavigation.getRoutes().first()
+            ) + firstAlternative.alternatives!! + externalAlternatives
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:227]: Is `firstAlternative.alternatives!!` something that was generated by NN? What do we expect to find in this list here?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 4 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:156]: the test checks that the observer is not triggered for `MapboxNavigation#setRoutes`, but waits for the very first trigger (that happens later). 

>Why do you need this check at all?

this check was added in the previous test (see `expect_initial_alternative_route_removed_after_navigating_past`), here we can re-use it for free without additional effort.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 5 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:254]: yes, it is. I wanted to verify logic, where `setRoutes` list is full enough (I don't need to control how many routes come to `AlternativesOberver`, I only want to check that it's full enought)

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 6 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:227]: We actually don't really worry about `firstAlternative.alternatives!!`, in this test we need to mix NN alternatives with the external route(s) and set the list as alternatives to `MapboxNavigation#setRoutes`. This way we check that NN is not triggering native AlternativeObserver

**Comment by RingerJK** [comment]
> > Alternatives observer does not spread routes that has been set via MapboxNavigation#setRoutes;
> > Alternatives observer does not spread routes if set to MapboxNavigation#setRoutesalternatives routes + one or few additional routes.
> 
> Do you mean that it was implemented by NN and you just removed the code that we don't need any longer?

That's what was changed on the NN side. We already handle the cases on the SDK side. So for the SDK we need to remove some logic and add tests which verify regression

File: CHANGELOG.md (hunk 2 above)
Comment (LukasPaczos) [reply CHANGELOG.md:39]: The string is still here 🙃

Comment (RingerJK) [reply CHANGELOG.md:39]: 🤔 it struggles!

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 5 above)
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:254]: Can we then check the size right after the creation? Here it's like hanging in the air. Quite confusing.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 6 above)
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:227]: What would happen if we set:
1) primary route + NN alternatives;
2) primary route + external alternatives
Would the logic be different? Do you think it's worth testing?
I thought I've left a similar comment somewhere but I can't find it...

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 4 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:156]: > check that the observer is triggered

could you point out what you mean?
```
        assertTrue(firstAlternative.calledOnMainThread)
```
and 
```
        assertNotNull(firstAlternative.routeProgress)
        assertNotNull(firstAlternative.alternatives)
```
checks are different

>wait until we stop recording (btw, what's recording have to do with the alternatives?)

the recoding is storing history recording files to profile tests if something goes wrong

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 6 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:227]: > primary route + NN alternatives;

that's what actually verifies at `alternative_routes_observer_is_not_triggered_for_set_routes` 

>primary route + external alternatives

it's actually the same scenario: NN had an issue with a set of alternatives that didn't provide by them. So excluding an alternative, removing the whole alternative list, or adding a new one to a list is the same scenario.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 5 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:254]: sure, I'll move checks.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 4 above)
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:156]: > check that the observer is triggered

I meant `assertTrue(firstAlternative.calledOnMainThread)`. It's only true when triggered. But OK, I see what you mean. You want to check that it's triggered but not for setRoutes. Then I just find the `calledOnMainThread` check redundant here.

> the recoding is storing history recording files to profile tests if something goes wrong

Oh, cool, I didn't use this in my tests because I wanted to keep them as simple as possible. BTW, can this be automised? I mean it would be cool if history was recorded for each test under-the-hood so that we don't overcomplicate the test body. WDYT? Should I cut a ticket for discussion or is it something that you've already discussed and decided it's not possible/we don't want that?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 7]
```diff
@@ -113,11 +116,150 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        countDownLatch.await()
+
+        // Verify alternative routes events were triggered.
+        assertNotNull(firstAlternative.routeProgress)
+        assertNotNull(firstAlternative.alternatives)
+
+        // Verify next alternative routes events were triggered without set routes
+        assertNotNull(nextAlternativeObserver.alternatives)
+        assertNotNull(nextAlternativeObserver.routeProgress)
+        // verify that set routes with  alternatives + 1 external alternatives is not triggered
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:256]: Why not?
`assertEquals(setRoute, nextAlternative)` implies that the routes passed via `setRoutes` are the same as the ones passed to the alternative observer, doesn't it?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 8]
```diff
@@ -113,11 +116,150 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        // alternative observer
+        nextAlternativeObserver.alternatives!!.forEach { nextAlternative ->
+            setRoutes.forEach { setRoute ->
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:259]: Wait, what's happening? You check that every route from setRoutes (there are >= 2 routes) is equal to one `nextAlternative` route?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 9]
```diff
@@ -113,11 +116,150 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+                countDownLatch.countDown()
+            }
+        }
+        countDownLatch.await()
+
+        // Verify alternative routes events were triggered.
+        assertNotNull(firstAlternative.routeProgress)
+        assertNotNull(firstAlternative.alternatives)
+
+        // Verify next alternative routes events were triggered without set routes
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:253]: How do these lines verify that it's not `setRoutes` that triggered the observer? I ask so many questions about the test because I'm still confused about it and foresee a situation in the future where I might need to change that. :)

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 7 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:256]: sorry, I see that it's a problem with tests. Need to fix: routes mustn't be the same! I need to cheks why the test passes 🤔

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 8 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:259]: https://github.com/mapbox/mapbox-navigation-android/pull/6214#discussion_r961784700

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 4 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:156]: >Then I just find the calledOnMainThread check redundant here.

I see what you mean 👍 
- don't see anything bad to verify threads in every place where it might be needed (we work with interfaces and we're agnostic to what is happening under an interface if switching to a new thread depends on any logic);
- I'd say we need a common approach to verify threads in tests. I think kind of rule of util which we might inject into a code block of instrumentation tests and verify threads. As we haven't had anything like this let's keep the same here and try to create something more generic. I cut the issue https://github.com/mapbox/navigation-sdks/issues/1970

> I mean it would be cool if history was recorded for each test under-the-hood so that we don't overcomplicate the test body. WDYT? Should I cut a ticket for discussion or is it something that you've already discussed and decided it's not possible/we don't want that?

yeah, agree. Great idea. Could you cut a ticket?

Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:156]: `navigation-sdks#1971`

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 10]
```diff
@@ -113,11 +116,150 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        assertTrue(firstAlternative.calledOnMainThread)
+
+        val countDownLatch = CountDownLatch(1)
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE("history path=$it", "DEBUG")
+                countDownLatch.countDown()
+            }
+        }
+        countDownLatch.await()
```
Comment (VysotskiVadim) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:165]: can it be extracted in a function? It seems to be repeated in many tests

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 11]
```diff
@@ -113,11 +116,150 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+
+        // Verify alternative routes events were triggered.
+        assertNotNull(firstAlternative.routeProgress)
+        assertNotNull(firstAlternative.alternatives)
```
Comment (VysotskiVadim) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:251]: TBH, I find this moment is quite tricky. Now I see the logic: if fields aren't nulls, it means that route alternative callback was called. But maybe it makes sense to incapsulate it in the `RouteAlternativesIdlingResource `? Like

```
class RouteAlternativesIdlingResource ... {
 ....

  fun verifyAlternativesCallbackWasTriggered() {
      assertNotNull(firstAlternative.routeProgress)
      assertNotNull(firstAlternative.alternatives)
  }
}
```
WDYT?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt [hunk 12]
```diff
@@ -52,14 +53,19 @@ class RouteAlternativesIdlingResource(
...
     ) {
+        if (waifForNonEmptyAlternativesList && alternatives.isEmpty()) {
+            mapboxNavigation.setNavigationRoutes(mapboxNavigation.getNavigationRoutes().take(1))
+            return
+        }
         // Verify this happens on the main thread.
-        if (Looper.myLooper() != Looper.getMainLooper()) {
-            calledOnMainThread = false
+        if (Looper.myLooper() == Looper.getMainLooper()) {
+            calledOnMainThread = true
```
Comment (VysotskiVadim) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt:62]: just curious, initial option didn't work for some case, did it?

File: instrumentation-tests/src/main/res/raw/route_response_alternative_continue.json [hunk 13]
```diff
@@ -0,0 +1,702 @@
+{
```
Comment (VysotskiVadim) [inline instrumentation-tests/src/main/res/raw/route_response_alternative_continue.json:1]: Is this file used anywhere?

Comment (VysotskiVadim) [reply instrumentation-tests/src/main/res/raw/route_response_alternative_continue.json:1]: found it

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 14]
```diff
@@ -113,11 +116,150 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+
+        // Wait for enhanced locations to start and then set the routes.
+        val firstLocationIdlingResource = FirstLocationIdlingResource(mapboxNavigation)
+        firstLocationIdlingResource.firstLocationSync()
+        runOnMainSync {
+            mapboxNavigation.setRoutes(primaryRouteOnly)
+        }
+
+        // The alternative route is missed, so we expect the route to be removed.
+        val firstAlternative = RouteAlternativesIdlingResource(mapboxNavigation, true)
```
Comment (VysotskiVadim) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:206]: how does NN get alternatives during the tests? As far as I understand it will wait until you passe the deviation point and then do a route request. But there is no setup responses for this case in mocked web server 🤔 
How does it work?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt (hunk 12 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt:62]: it worked, make it reverted to avoid check passed if `calledOnMainThread` is not called.  Ideally there should be an additional check something like `wasInvoked`, but it's enough for now I think

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 10 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt]: I think it's good to refactor, let's address it separately

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 14 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt]: it works independently, so alternatives are produced by NN for the current mocked route

Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt]: The test is agnostic to the type of route, we don't need to know it. But NN might produce routes by onboard or offboard, depending on current conditions (like internet connection availability).

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt [hunk 15]
```diff
@@ -52,14 +58,24 @@ class RouteAlternativesIdlingResource(
         alternatives: List<DirectionsRoute>,
         routerOrigin: RouterOrigin
     ) {
+        if (!onRoutesAlternatives.invoke(routeProgress, alternatives, routerOrigin)) {
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt:61]: It's not really clear from the name `onRoutesAlternatives` when it returns true and when false. Because of that it's also not clear why you return here. Could you specify that in the name or at least in the docs?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt [hunk 16]
```diff
@@ -52,14 +58,24 @@ class RouteAlternativesIdlingResource(
...
         }
 
         mapboxNavigation.unregisterRouteAlternativesObserver(this)
         this.routeProgress = routeProgress
         this.alternatives = alternatives
+
         callback?.onTransitionToIdle()
     }
+
+    fun verifyOnRouteAlternativesPassed() {
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/idling/RouteAlternativesIdlingResource.kt:77]: What does "passed" mean here? I'd think it's something like "the vehicle was following this route". But it seems like it's "alternative routes were received via callback". Can we change that to `verifyOnRouteAlternativesReceived` or something similar?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 17]
```diff
@@ -107,7 +119,58 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        }
+    }
+
+    @Test
+    fun alternative_routes_observer_is_not_triggered_for_set_routes() {
+        // Prepare with alternative routes.
+        setupMockRequestHandlers(coordinates)
+        val routes = requestDirectionsRouteSync(coordinates)
+
+        // Play the slower alternative route.
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:143]: Why the alternative? `routes.first()` means primary, doesn't it?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 18]
```diff
@@ -107,7 +119,58 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        assertEquals(2, routes.size)
+        assertTrue(mapboxNavigation.getNavigationRoutes().isNotEmpty())
+        firstAlternative.verifyOnRouteAlternativesPassed()
+        firstAlternative.alternatives!!.forEach {
+            assertFalse(routes.contains(it))
+        }
+    }
+
+    @Test
+    fun alternative_routes_observer_is_not_triggered_for_set_routes() {
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:138]: Where is it checked that it's not triggered? I only see `firstAlternative.verifyOnRouteAlternativesPassed()` which means it was actually triggered.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 19]
```diff
@@ -107,7 +119,58 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        }
+
+        // Wait for enhanced locations to start and then set the routes.
+        val firstLocationIdlingResource = FirstLocationIdlingResource(mapboxNavigation)
+        firstLocationIdlingResource.firstLocationSync()
+        runOnMainSync {
+            mapboxNavigation.setRoutes(routes)
+        }
+
+        // The alternative route is missed, so we expect the route to be removed.
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:157]: Which route? Removed from where? And where is it checked? I'm confused.
Is it like the next test but for for the routes that were set? Then why don't you check the lists for inequality like there?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 20]
```diff
@@ -116,8 +179,121 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+        ) { _, alternatives, _ ->
+            alternatives.isNotEmpty()
+        }
+
+        val externalAlternatives = DirectionsResponse.fromJson(
+            readRawFileText(activity, R.raw.route_response_alternative_continue)
+        ).routes().also {
+            assertEquals(1, it.size)
+        }
+        var setRoutes: List<DirectionsRoute> = emptyList()
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:253]: lateinit var?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 18 above)
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:138]: I noticed that you use deprecated `setRoutes`/`getRoutes`. I think it would be a good idea to use the actual API in new tests. Old ones can be left as is while the old API works but why increase its usage?

Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:138]: it must be triggered, we check that alternatives that come in `AlternativesObserver` are different than we set in `setRoutes`

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 19 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:157]: changed the description

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 18 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:138]: it's needed to respect `RouteAlternativesIdlingResource` which handles `DirectionsRoute`s. I would change the whole flow to `NavigationRoutes`, let's do it separately

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 14 above)
Comment (VysotskiVadim) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt]: Sorry, didn't get this 🙂 
Let's discuss on a call later

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 18 above)
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:138]: Okay

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 21]
```diff
@@ -116,8 +179,121 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+                ).also {
+                assertTrue(
+                    "Primary route + >=1 alternatives + external alternatives",
+                    it.size >= 3
+                )
+            }
+            mapboxNavigation.setRoutes(setRoutes)
+        }
+
+        nextAlternativeObserver.register()
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:268]: Ir seems like RoutesAlternativeObserver is not sticky. I think a race is possible here: alternative observers will be notified before this one is registered.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 22]
```diff
@@ -107,7 +119,58 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
+            mapboxNavigation.setRoutes(routes)
+        }
+
+        // Subscribing for alternatives
+        val firstAlternative = RouteAlternativesIdlingResource(
+            mapboxNavigation
+        ) { _, alternatives, _ ->
+            alternatives.isNotEmpty()
+        }
+        firstAlternative.register()
```
Comment (LukasPaczos) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:163]: This should be done before
```
mapboxNavigation.setRoutes(routes)
```

Otherwise, if the alternatives observer was caught synchronously, we wouldn't catch it.

Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:163]: make sense

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 21 above)
Comment (RingerJK) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:268]: that's actually mustn't, as I added tests for that `alternative_routes_observer_is_not_triggered_for_set_routes` (set route is not trigger observer), but I'll move `nextAlternativeObserver.register()` above 👍

**Review by dzinad — APPROVED** [review]
